src/Build/Evaluation/LazyItemEvaluator.LazyItemOperation.cs

```diff
@@ -110,14 +110,14 @@ protected void DecorateItemsWithMetadata(ICollection<I> items, ImmutableList<Par
 
                     // Do not expand properties as they have been already expanded by the lazy evaluator upon item operation construction.
                     // Prior to lazy evaluation ExpanderOptions.ExpandAll was used.
-                    const ExpanderOptions metadataExpansionOptions = ExpanderOptions.ExpandMetadata | ExpanderOptions.ExpandItems;
```

Comment:
Good catch @radical!

After an adventurous debugging session, it seems that `%(RelativeDir)` does not get expanded at operation application because [the result of a property function is always escaped](https://github.com/Microsoft/msbuild/blob/master/src/Build/Evaluation/Expander.cs#L3167-L3173). This means that `$([System.String]::new('%(RelativeDir)').TrimEnd('\'))` expands into `%25%28RelativeDir%29` which is then not detected as metadata during operation application.

The before-the-regression code and your fix work because if metadata expansion is delayed until operation application, it will get expanded with `ExpanderOptions.ExpandAll` which makes the `Expander` expand references in a [specific order](https://github.com/Microsoft/msbuild/blob/master/src/Build/Evaluation/Expander.cs#L253-L255): 1) metadata, 2) properties, 3) items. Since metadata is expanded before properties, the property function gets the correct values.

Consequences:
```xml
<Project>
        
        <ItemGroup>
            <A Include="1;$(P)" Exclude="$(P)"/>
            <B Include="a\**">
              <WithMetArg>$([System.String]::new('%(RelativeDir)').TrimEnd('\'))</WithMetArg>
              <WithIndirection>$(P)</WithIndirection>
            </B>
            <C Include="$([System.String]::new('@(A)'))"/>
            <D Include="$([System.String]::new('$(P)'))"/>
        </ItemGroup>

        <PropertyGroup>
            <P>@(A)</P>
        </PropertyGroup>

        <Target Name="Main">
                <Message Text="WithMetArg: %(B.WithMetArg)"/>
                <Message Text="WithIndirection: %(B.WithIndirection)"/>
                <Message Text="Item as property function argument: @(C)"/>
                <Message Text="Property as property function argument: @(D)"/>
        </Target>
</Project>
```

Evaluates to the following on dev14
```
Microsoft (R) Build Engine version 14.0.25420.1
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 4/5/2017 1:39:25 PM.
Project "D:\projects\tests\projects\play\build.proj" on node 1 (default targets).
Main:
  WithMetArg: a\b
  WithIndirection: 1
  Item as property function argument: @(A)
  Property as property function argument: @(A)
Done Building Project "D:\projects\tests\projects\play\build.proj" (default targets).
```

Notice that the property reference inside the argument does get expanded, because arguments are actually expanded with the same `ExpanderOptions` as the initial string: https://github.com/Microsoft/msbuild/blob/master/src/Build/Evaluation/Expander.cs#L3050

I'll add tests to capture this behaviour.